api/src/controllers/validation.ts

```diff
@@ -1,51 +1,92 @@
+type NonEmptyArray<T> = [T, ...T[]];
 
 type QueryValidatorConfig<Name, AllowedValue> = {
   name: StringLiteral<Name>;
-  defaultValue: Readonly<StringLiteral<AllowedValue>>;
   allowed: ReadonlyArray<StringLiteral<AllowedValue>>;
+  defaultValue?: Readonly<StringLiteral<AllowedValue>>;
+  singleValue?: boolean;
 };
-
-export const queryValidator =
-  <Name, AllowedValue>({
-    name,
-    defaultValue,
-    allowed,
-  }: QueryValidatorConfig<Name, AllowedValue>) =>
-  <Query extends { [key in typeof name]?: AllowedValue | string }>(
+export const queryValidator = <Name, AllowedValue>({
+  name,
+  allowed,
+  defaultValue,
+  singleValue,
+}: QueryValidatorConfig<Name, AllowedValue>) => {
+  const allowedSet = new Set(allowed);
+  return <Query extends { [key in typeof name]?: string }>(
     query: Query
-  ): AllowedValue => {
-    const value: AllowedValue | string | undefined = query[name];
-    if (typeof value === "undefined") {
-      return defaultValue;
-    } else if (isInArray(value, allowed)) {
-      return value;
-    } else {
+  ): NonEmptyArray<AllowedValue> | undefined => {
+    const providedValues = query[name]?.split(",");
+    if (providedValues === undefined) {
+      return defaultValue === undefined ? undefined : [defaultValue];
+    }
+
+    const validValues = providedValues.filter(isInSet(allowedSet));
+    if (
+      validValues.length !== providedValues.length ||
+      validValues.length === 0
+    ) {
+      const invalidValues = providedValues.filter(not(isInSet(allowedSet)));
+      const invalidMessage =
+        invalidValues.length === 1
+          ? "is not a valid value"
+          : "are not valid values";
       throw new HttpError({
         status: 400,
         label: "Bad Request",
-        description:
-          `${name}: '${value}' is not a valid value. Please choose one of: ` +
-          allowed.map((s) => `'${s}'`).join(", "),
+        description: `${name}: ${readableList(
+          invalidValues
+        )} ${invalidMessage}. Please choose one of ${readableList(
+          allowed,
+          "or"
+        )}`,
       });
     }
+
+    if (singleValue && validValues.length > 1) {
+      throw new HttpError({
+        status: 400,
+        label: "Bad Request",
+        description: `Only 1 value can be specified for ${name}`,
+      });
+    }
+
+    return validValues as NonEmptyArray<AllowedValue>;
   };
+};
 
 export const validateDate = (input: string): Date => {
   const date = new Date(input);
   if (isNaN(date.getTime())) {
     throw new HttpError({
       status: 400,
       label: "Bad Request",
-      description: `'${input}' is not a valid date. Please specify a date or datetime in ISO 8601 format.`,
+      description: `${quoted(
+        input
+      )} is not a valid date. Please specify a date or datetime in ISO 8601 format.`,
     });
   }
   return date;
 };
+
+const quoted = (str: string) => `'${str}'`;
+
+// For listing items (no Oxford comma)
+const readableList = (
```

Comment:
nit: not sure if this belongs in a controllers file or would be better pushed into some sort of grammar utils file (cf the front-end apps)

Reply:
At some point I predict that these utils will be commonised (ie `typescript-libs`) so will deal with that then